adapters/huaweiads/huaweiads.go

```diff
@@ -0,0 +1,1320 @@
+	}
+	var totalSec = int(duration) / 1000
+	var mmm = int(duration) % 1000
+	var hour = totalSec / 3600
+	var tmp = totalSec % 3600
+	var min = tmp / 60
+	var sec = tmp % 60
+	var result string
+	result += addZero(strconv.Itoa(hour), ":", false)
+	result += addZero(strconv.Itoa(min), ":", false)
+	result += addZero(strconv.Itoa(sec), ".", false)
+	result += addZero(strconv.Itoa(mmm), "", true)
+	return result
+}
+
+func addZero(str string, dot string, isMmm bool) (result string) {
+	if isMmm == false {
+		if len(str) < 2 {
+			result = "0" + str + dot
+		} else {
+			result = str + dot
+		}
+	} else {
+		if len(str) == 1 {
+			result = "00" + str
+		} else if len(str) == 2 {
+			result = "0" + str
+		} else {
+			result = str
+		}
+	}
+	return result
+}
+
+// get field adm for video, vast 3.0
+func (a *adapter) extractAdmVideo(adType int32, content *content, bidType openrtb_ext.BidType, opentrb2Imp openrtb2.Imp) (adm string,
+	adWidth int64, adHeight int64, err error) {
+	if content == nil {
+		return "", 0, 0, fmt.Errorf("extractAdmVideo: content is empty")
+	}
+
+	var clickUrl = ""
+	if content.MetaData.ClickUrl != "" {
+		clickUrl = content.MetaData.ClickUrl
+	} else if content.MetaData.Intent != "" {
+		clickUrl, _ = getDecodeValue(content.MetaData.Intent)
+	}
+
+	var mime = "video/mp4"
+	var resourceUrl = ""
+	var duration = ""
+	if adType == roll {
+		// roll ad get information from mediafile
+		if content.MetaData.MediaFile.Mime != "" {
+			mime = content.MetaData.MediaFile.Mime
+		}
+		adWidth = content.MetaData.MediaFile.Width
+		adHeight = content.MetaData.MediaFile.Height
+		if content.MetaData.MediaFile.Url != "" {
+			resourceUrl = content.MetaData.MediaFile.Url
+		} else {
+			return "", 0, 0, errors.New("extractAdmVideo: Content.MetaData.MediaFile.Url is empty")
+		}
+		duration = getDuration(content.MetaData.Duration)
+	} else {
+		if content.MetaData.VideoInfo.VideoDownloadUrl != "" {
+			resourceUrl = content.MetaData.VideoInfo.VideoDownloadUrl
+		} else {
+			return "", 0, 0, errors.New("extractAdmVideo: content.MetaData.VideoInfo.VideoDownloadUrl is empty")
+		}
+		if content.MetaData.VideoInfo.Width != 0 && content.MetaData.VideoInfo.Height != 0 {
+			adWidth = int64(content.MetaData.VideoInfo.Width)
+			adHeight = int64(content.MetaData.VideoInfo.Height)
+		} else if bidType == openrtb_ext.BidTypeVideo {
+			if opentrb2Imp.Video != nil && opentrb2Imp.Video.W != 0 && opentrb2Imp.Video.H != 0 {
+				adWidth = opentrb2Imp.Video.W
+				adHeight = opentrb2Imp.Video.H
+			}
+		} else {
+			return "", 0, 0, errors.New("extractAdmVideo: cannot get width, height")
+		}
+		duration = getDuration(int64(content.MetaData.VideoInfo.VideoDuration))
+	}
+
+	var adTitle, _ = getDecodeValue(content.MetaData.Title)
+	var adId = content.Contentid
+	var creativeId = content.Contentid
+	var trackingEvents strings.Builder
+	var dspImpTracking2Str = ""
+	var dspClickTracking2Str = ""
+	var errorTracking2Str = ""
+	if content.Monitor != nil {
```

Comment:
I don't believe the nil check is necessary. The range loop gracefully handles nil slices.

Reply:
corrected